flows/actions/reply.go

```diff
@@ -43,12 +49,55 @@ func (a *ReplyAction) Validate(assets flows.SessionAssets) error {
 func (a *ReplyAction) Execute(run flows.FlowRun, step flows.Step, log flows.EventLog) error {
 	evaluatedText, evaluatedAttachments, evaluatedQuickReplies := a.evaluateMessage(run, step, a.Text, a.Attachments, a.QuickReplies, log)
 
-	contactURNs := run.Contact().URNs()
+	channelSet, err := run.Session().Assets().GetChannelSet()
+	if err != nil {
+		return err
+	}
+
+	channels := channelSet.WithRole(flows.ChannelRoleSend)
+	destinations := []msgDestination{}
 
-	if a.AllURNs && len(contactURNs) > 0 {
-		log.Add(events.NewBroadcastCreatedEvent(evaluatedText, evaluatedAttachments, evaluatedQuickReplies, contactURNs.RawURNs(false), nil, nil))
+	if a.AllURNs {
+		// send to any URN which has a corresponding channel (i.e. is sendable)
+		for _, u := range run.Contact().URNs() {
+			channel := getChannelForURN(channels, u)
+			if channel != nil {
+				destinations = append(destinations, msgDestination{urn: u.URN, channel: channel})
+			}
+		}
 	} else {
-		log.Add(events.NewMsgCreatedEvent(evaluatedText, evaluatedAttachments, evaluatedQuickReplies, run.Contact().Reference()))
+		// send to first URN which has a corresponding channel (i.e. is sendable)
+		for _, u := range run.Contact().URNs() {
+			channel := getChannelForURN(channels, u)
+			if channel != nil {
+				destinations = append(destinations, msgDestination{urn: u.URN, channel: channel})
+				break
+			}
+		}
```

Comment:
These loops look the same except you could do the break conditionally based on `allURNs`